File: src/integrationTest/java/net/ltgt/gradle/errorprone/GroovyDslIntegrationTest.java
package net.ltgt.gradle.errorprone;

import static com.google.common.truth.Truth.assertThat;
import static java.util.Objects.requireNonNull;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import org.gradle.testkit.runner.TaskOutcome;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class GroovyDslIntegrationTest extends BasePluginIntegrationTest {
  @Override
  protected Path getSettingsFile() {
    return projectDir.resolve("settings.gradle");
  }

  @Override
  protected Path getBuildFile() {
    return projectDir.resolve("build.gradle");
  }

  @BeforeEach
  void setup() throws Exception {
    Files.writeString(
        getBuildFile(),
        // language=groovy
        """
        plugins {
            id("java-library")
            id("%s")
        }
        repositories {
            mavenCentral()
        }
        dependencies {
            errorprone "com.google.errorprone:error_prone_core:%s"
        }
        """
            .formatted(ErrorPronePlugin.PLUGIN_ID, errorproneVersion));
  }

  @Test
  void canDisableErrorProne() throws Exception {
    Files.writeString(
        getBuildFile(),
        // language=groovy
        """

        tasks.withType(JavaCompile).configureEach {
            options.errorprone.enabled = false
        }
        """,
        StandardOpenOption.APPEND);
    writeFailureSource();

    // when
    var result = buildWithArgs("compileJava");

    // then
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
  }

  @Test
  void canConfigureErrorProne() throws Exception {
    // given
    Files.writeString(
        getBuildFile(),
        // language=groovy
        """

        import net.ltgt.gradle.errorprone.CheckSeverity
        import kotlin.Pair

        tasks.withType(JavaCompile).configureEach {
            options.errorprone {
                // configure with the default values, we only want to check the DSL, not the effects
                enabled = true
                disableAllChecks = false
                disableAllWarnings = false
                allErrorsAsWarnings = false
                allSuggestionsAsWarnings = false
                allDisabledChecksAsWarnings = false
                disableWarningsInGeneratedCode = false
                ignoreUnknownCheckNames = false
                ignoreSuppressionAnnotations = false
                compilingTestOnlyCode = false
                excludedPaths = "should.not.match.anything"

                check new Pair("Foo", CheckSeverity.ERROR), new Pair("Bar", CheckSeverity.DEFAULT)
                check "Foo", CheckSeverity.WARN
                check "Bar", provider { CheckSeverity.OFF }
                enable "Foo", "Bar"
                disable "Foo", "Bar"
                warn "Foo", "Bar"
                error "Foo", "Bar"
                checks.empty()

                option "Foo:Bar"
                option "Foo:Bar", false
                option "Foo:Bar", "baz"
                option "Foo:Bar", provider { "baz" }
                checkOptions.empty()

                errorproneArgs.empty()
                errorproneArgumentProviders.clear()
            }
        }
        """,
        StandardOpenOption.APPEND);
    writeSuccessSource();

    // when
    var result = buildWithArgs("compileJava");

    // then
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
  }
}

File: src/integrationTest/kotlin/net/ltgt/gradle/errorprone/GroovyDslIntegrationTest.kt
package net.ltgt.gradle.errorprone

import com.google.common.truth.Truth.assertThat
import org.gradle.testkit.runner.TaskOutcome
import org.gradle.util.GradleVersion
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.util.Properties

class GroovyDslIntegrationTest {
    @TempDir
    lateinit var testProjectDir: File
    lateinit var settingsFile: File
    lateinit var buildFile: File

    @BeforeEach
    fun setupProject() {
        assumeCompatibleGradleAndJavaVersions()

        testProjectDir.resolve("gradle.properties").outputStream().use {
            Properties().apply {
                setProperty("org.gradle.java.home", testJavaHome)
                store(it, null)
            }
        }
        settingsFile =
            testProjectDir.resolve("settings.gradle").apply {
                createNewFile()
            }
        buildFile =
            testProjectDir.resolve("build.gradle").apply {
                appendText(
                    """
                    plugins {
                        id("java-library")
                        id("${ErrorPronePlugin.PLUGIN_ID}")
                    }
                    repositories {
                        mavenCentral()
                    }
                    dependencies {
                        errorprone "com.google.errorprone:error_prone_core:$errorproneVersion"
                    }
                    """.trimIndent(),
                )
            }
        if (testGradleVersion < GradleVersion.version("7.0")) {
            buildFile.appendText(
                """

                allprojects {
                    configurations.all {
                        attributes.attribute(Attribute.of("org.gradle.jvm.environment", String), "standard-jvm")
                    }
                }
                """.trimIndent(),
            )
        }
    }

    @Test
    fun `can disable errorprone`() {
        // given
        buildFile.appendText(
            """

            tasks.withType(JavaCompile).configureEach {
                options.errorprone.enabled = false
            }
            """.trimIndent(),
        )
        testProjectDir.writeFailureSource()

        // when
        val result = testProjectDir.buildWithArgs("compileJava")

        // then
        assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
    }

    @Test
    fun `can configure errorprone`() {
        // given
        buildFile.appendText(
            """

            tasks.withType(JavaCompile).configureEach {
                options.errorprone {
                    // configure with the default values, we only want to check the DSL, not the effects
                    enabled = true
                    disableAllChecks = false
                    disableAllWarnings = false
                    allErrorsAsWarnings = false
                    allDisabledChecksAsWarnings = false
                    disableWarningsInGeneratedCode = false
                    ignoreUnknownCheckNames = false
                    ignoreSuppressionAnnotations = false
                    compilingTestOnlyCode = false
                    excludedPaths = "should.not.match.anything"
                    enable("ArrayEquals")
                }
            }
            """.trimIndent(),
        )
        testProjectDir.writeSuccessSource()

        // when
        val result = testProjectDir.buildWithArgs("compileJava")

        // then
        assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
    }
}
